Fix CVE-2019-12995 blog post (#4584)
istio/istio.io: content/blog/2019/cve-2019-12995/index.md

---
title: Security Update - CVE-2019-12995
description: Security vulnerability disclosure for CVE-2019-12995.
publishdate: 2019-06-28

> attribution: The Istio Team

Reviewer: We should probably use a keyword such as CVE on this and other CVE blog post for easy searching and categorization.

Recommend adding the tag

`keywords: [CVE]`

Author: ```suggestion
attribution: The Istio Team
keywords: [CVE]
```

---

A bug in Istio’s JWT validation filter causes Envoy to crash in certain cases when the request contains a malformed JWT token. The bug was discovered and reported by a user [on GitHub](https://github.com/istio/istio/issues/15084) on June 23, 2019.

This bug affects all versions of Istio that are using the JWT authentication policy.

The symptoms of the bug are an HTTP 503 error seen by the client, and

{{< text plain >}}
Epoch 0 terminated with an error: signal: segmentation fault (core dumped)
{{< /text >}}

in the Envoy logs.

> The Envoy crash can be triggered using a malformed JWT without a valid signature, and on any URI being accessed regardless of the `trigger_rules` in the JWT specification. Thus, this bug makes the Envoy vulnerable to a potential DoS attack.

Author: ```suggestion
The Envoy crash can be triggered using a malformed JWT without a valid signature, and on any URI being accessed regardless of the `trigger_rules` in the JWT specification. Thus, this bug makes Envoy vulnerable to a potential DoS attack.
```

## Vulnerability impact and Detection

Envoy is vulnerable if the following two conditions are satisfied:
* A JWT authentication policy is applied to it.

> * The JWT issuer (specified by `jwksUri`) uses the RSA algorithm for signature verification (Note the bug is just triggered in this condition and has nothing to do with the RSA algorithm)

Reviewer: we strongly strive not to use the word note.  If we see the word note, we generally prefer a callout: https://istio.io/about/contribute/creating-and-editing-pages/#callouts

I'd recommend something like
```
{{< tip >}}
The RSA algorithm used for signature verification does not contain any known security vulnerability.  This CVE is triggered only when using this algorithm but is unrelated to the security of the system.
<< /tip >}}
```

Author: ```suggestion
* The JWT issuer (specified by `jwksUri`) uses the RSA algorithm for signature verification

{{< tip >}}
The RSA algorithm used for signature verification does not contain any known security vulnerability.  This CVE is triggered only when using this algorithm but is unrelated to the security of the system.
<< /tip >}}

```

Reviewer: line 44 should begin with {. (sorry had typo).

If JWT policy is applied to the Istio ingress gateway, please be aware that any external user who has access to the ingress gateway could crash it with a single HTTP request.

If JWT policy is applied to the sidecar only, please keep in mind it might still be vulnerable. For example, the Istio ingress gateway might forward the JWT token to the sidecar which could be a malformed JWT token that crashes the sidecar.

> A vulnerable Envoy will crash on an HTTP request with a malformed JWT token. When Envoy crashes, all existing connections will be disconnected immediately. The pilot-agent will restart the crashed Envoy automatically and it may take a few seconds to a few minutes for the restart. pilot-agent will stop restarting Envoy after it crashed more than 10 times. In this case, Kubernetes will redeploy the pod (including the workload behind Envoy).

Reviewer: I suspect pilot-agent needs to be in ```.  Spell out the number 10 (use ten instead).

Remove the trailing `)` as there is no opening `(`).

Author: ```suggestion
A vulnerable Envoy will crash on an HTTP request with a malformed JWT token. When Envoy crashes, all existing connections will be disconnected immediately. The `pilot-agent` will restart the crashed Envoy automatically and it may take a few seconds to a few minutes for the restart. pilot-agent will stop restarting Envoy after it crashed more than ten times. In this case, Kubernetes will redeploy the pod, including the workload behind Envoy.
```

Envoy is vulnerable if the following two conditions are satisfied:
* A JWT authentication policy is applied to it.
* The JWT issuer (specified by `jwksUri`) uses the RSA algorithm for signature verification

{{< tip >}}
The RSA algorithm used for signature verification does not contain any known security vulnerability.  This CVE is triggered only when using this algorithm but is unrelated to the security of the system.

> << /tip >}}

Reviewer: needs to be `{{<` I indicated the wrong change earlier apologies.

Author: ```suggestion
{{< /tip >}}
```
